crates/sui-data-ingestion/src/workers/archival.rs

```diff
@@ -0,0 +1,200 @@
+            commit_duration: Duration::from_secs(config.commit_duration_seconds),
+        })
+    }
+
+    async fn upload(&self, state: &AccumulatedState) -> Result<()> {
+        let checkpoint_file_path =
+            format!("epoch_{}/{}.chk", state.epoch, state.checkpoint_range.start);
+        self.upload_file(
+            Path::from(checkpoint_file_path.clone()),
+            CHECKPOINT_FILE_MAGIC,
+            &state.buffer,
+        )
+        .await?;
+        let summary_file_path =
+            format!("epoch_{}/{}.sum", state.epoch, state.checkpoint_range.start);
+        self.upload_file(
+            Path::from(summary_file_path.clone()),
+            SUMMARY_FILE_MAGIC,
+            &state.summary_buffer,
+        )
+        .await?;
+        let manifest_bytes = self
+            .remote_store
+            .get(&Path::from("MANIFEST"))
+            .await?
+            .bytes()
+            .await?;
+        let mut manifest = read_manifest_from_bytes(manifest_bytes.to_vec())?;
+        let checkpoint_file_metadata = create_file_metadata(
+            std::path::Path::new(&checkpoint_file_path),
+            FileType::CheckpointContent,
+            state.epoch,
+            state.checkpoint_range.clone(),
+        )?;
+        let summary_file_metadata = create_file_metadata(
+            std::path::Path::new(&summary_file_path),
+            FileType::CheckpointSummary,
+            state.epoch,
+            state.checkpoint_range.clone(),
+        )?;
+        manifest.update(
+            state.epoch,
+            state.checkpoint_range.start,
+            checkpoint_file_metadata,
+            summary_file_metadata,
+        );
+        let bytes = finalize_manifest(manifest)?;
+        self.remote_store
+            .put(&Path::from("MANIFEST"), bytes)
+            .await?;
+        Ok(())
+    }
+
+    async fn upload_file(&self, location: Path, magic: u32, content: &[u8]) -> Result<()> {
+        let mut buffer = vec![0; 4];
+        BigEndian::write_u32(&mut buffer, magic);
+        buffer.push(StorageFormat::Blob.into());
+        buffer.push(FileCompression::Zstd.into());
+        buffer.extend_from_slice(content);
+        let mut compressed_buffer = vec![];
+        let mut cursor = Cursor::new(buffer);
+        compress(&mut cursor, &mut compressed_buffer)?;
+        self.remote_store
+            .put(&location, Bytes::from(compressed_buffer))
+            .await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Worker for ArchivalWorker {
+    async fn process_checkpoint(&self, checkpoint: CheckpointData) -> Result<()> {
+        let mut state = self.state.lock().await;
+        let sequence_number = checkpoint.checkpoint_summary.sequence_number;
+        if sequence_number < state.checkpoint_range.start {
+            return Ok(());
+        }
+        let epoch = checkpoint.checkpoint_summary.epoch;
+        if state.checkpoint_range.is_empty() {
+            state.reset(epoch, sequence_number);
+        }
```

Comment:
just curious - do we still need this if condition? It looks like we invoke `reset` after upload and in the constructor which should be enough

Reply:
yes, it looks redundant now. Removed it. But please double check the constructor

Reply:
Constructor looks ok to me, but could be add a small test like this to get more confidence if possible https://github.com/MystenLabs/sui/blob/5b2f41bb33922d2d45be35db7a3be6626eefa7a1/crates/sui-archival/src/tests.rs#L197

Reply:
adding similar test would be hard at this point, storage is not abstracted away, we can do it as a followup or figure out in general how to test such changes.
Meanwhile, please check https://github.com/MystenLabs/sui/pull/16085/commits/f345267686fb6257263e0197609c6f613a773193#diff-4ed63633494b798c8de136ac9ea0141a04dc80025339cde44361b6929dddbe35R114
It's a similar test(not meant for landing) that uploads checkpoints to temporary S3 bucket and runs verification with the reader